
@main("Help", "MIMP", "Help") {

  <h4>What is MIMP?</h4>
  <div id="1">
    <p>Mutations IMpact on Phosphorylation (MIMP) is a web service that predicts if a mutation within a phosphorylation site will alter phosphorylation. MIMP works by constructing specificity models for kinases, which are then used to score phosphosites containing a mutation before and after the mutation to predict the impact it may have on phosphorylation. There are three types of events that can occur</p>
    <ul>
    <li>Loss-of-phosphorylation: the mutation disrupts specificity for a kinase</li>
    <li>Gain-of-phosphorylation: the mutation creates a new motif which causes phosphorylation by a new kinase</li>
    </ul>

    <img src="@routes.Assets.at("images/fig1.png")" style="height:400px">
  </div>

  <hr>

  <h4>How do I use it?</h4>
  <div id="2">
    <p>To use MIMP you will need two mandatory files and one optional file. </p>
    
    <ol>
    	<li>FASTA file of protein sequences. For example: 
    		<pre>>seq0
FQTWEEFSRAAEKLYLADPMKVRVVLKYRHVDGNLCIKVTDDLVCLVYRTDQAQDVKKIEKF
>seq1
KYRTWEEFTRAAEKLYQADPMKVRVVLKYRHCDGNLCIKVTDDVVCLLYRTDQAQDVKKIEKFHSQLMRLME LKVTDNKECLKFKTDQAQEAKKMEKLNNIFFTLM</pre>
    		<p><em>Note:</em> Avoid using spaces in your fasta headers. If spaces are used, MIMP will use the text before the first space as the header. For example, <code>>seq more text</code> will become <code>>seq</code>. If you do use spaces in your headers, make sure your mutation and phosphorylation data correspond to the equivalent of <code>>seq</code> in this case.</p>
    	</li>
    	<li>Mutation data. This should contain the name of the protein and the mutation, space delimited. The mutation should have a <code>X123Z</code> format, where <code>X</code> is the reference amino acid, <code>Z</code> is the alternate amino acid and <code>123</code> is the position of the substitution. Protein names should correspond to that of the fasta file. For example:
    	<pre>seq0 F1H
seq1 T4V
seq1 W5L</pre>
    	
    	</li>
    	<li>Phosphorylation site data <b>(optional but recommended)</b>. This should contain the name of the protein and the position of phosphoacceptor S, T or Y residue, space delimited. Protein names should correspond to that of the fasta file. For example:
    	<pre>seq0 8
seq0 15
seq1 9</pre>
		All positions must correspond to an S, T or Y residue. If you are not providing phosphorylation site data, MIMP will consider each S, T or Y residue as a potential phosphoacceptor.
    	</li>
    <ol>
  </div>


  <hr>
  
  
  <h4>What are the different kinase specificity models?</h4>
  <p>There are three different sets of specificity models. The chosen set is used in the scanning of phosphorylation rewiring in MIMP. </p>
  <ol>
	<li><b>Individual kinase specificity models</b> these are specificity models constructed using experimentally verified kinase-peptide phosphorylation on an individual kinase level. Data is obtained from public databases such as PhosphoSitePlus, PhosphoELM and HPRD. Since these models are constructed using experimental data on the peptide level, they are considered high confidence.</li>
	<li><b>Kinase family specificity models</b> This is similar to 1. Since kinase families typically have similar specificities, these models are also constructed using experimentally verified kinase-peptide phosphorylation, but are grouped based on the kinase families/subfamilies. This can have certain advantages and disadvantages. For example, there are cases where members of kinase families show a moderate and sometimes large difference from that of the collapsed family model. However, using kinase family models can avoid redundant (gain or loss of phosphorylation) prediction of family members that have similar specificities.</li>
	<li><b>Predicted individual kinase specificity models</b> these are specificity models that are predicted from kinase-substrate interactions, on the protein level and not the peptide. The kinase-substrate relationships (KSRs) are derived using protein microarrays as described in <a href="http://www.ncbi.nlm.nih.gov/pubmed/23549483" target="_blank">Newman <em>et al.</em> (2013)</a>. The KSRs are then combined with experimentally identified phosphosites to predict the specificity models by through an iterative refinement process (termed M3) similar to that described in the MIMP paper. For more information on the M3 algorithm please see the <a href="http://www.ncbi.nlm.nih.gov/pubmed/23549483" target="_blank">Newman <em>et al.</em> (2013)</a> paper.
	<br>
	<p class="warning-link">Please be extremely cautious when using these models. While KSRs are identified <em>in vitro</em> and some models may be correct, it is likely that others are not. These models are thus considered low-confidence and we strongly urge users who use this dataset to manually verify that the logos appear valid before proceeding with downstream analyses. </p>
	
	</li>
  </ol>
	

  <hr>
  
  <h4>What do the scores mean?</h4>
  <p>The scores quantitatively represent how well a phosphosite is likely to be phosphorylated by a kinase. Scores range from 0-1, where 0 represents no predicted phosphorylation and 1 represents perfect phosphorylation. 
  <br>Peptides are automatically given a score of 0 if they have a central residue which does not match the class of the specificity model it is being scored against. There are two classes of specifcity models: serine/threonine models and tyrosine models. For example, if the tyrosine-centered phosphosite <code>HAWNKDR<u>Y</u>QIAICPN</code> is scored against a serine/threonine model, it will always have a score of 0. If a peptide is mutated at the central residue to a non-STY residue, e.g. <code>HAWNKDR<u>A</u>QIAICPN</code> it will always have a score of 0.</p>

  <hr>
  
  <h4>What is the probability threshold?</h4>
  <p>Once a score is obtained for both the wild type and mutant phosphosite, MIMP will use kinase specificity models to score phosphosites that belong to a kinase (positive) and that do not (negative). This results in two distributions being generated: foreground (FG) and background (BG). 
  MIMP then computes a probability of the wild type or mutant score falling under either one of those distributions.  
  </p>
  <p>
	The probability of a phosphorylation loss is computed as the joint probability of the wild type score falling under the FG and the mutant score falling under the BG. 
	Similarly, the probability of a phosphorylation gain is computed as the joint probability of the wild type score falling under the BG and the mutant score falling under the FG. 
  </p>
	
  <p>
  	Since the joint probability is computed as the product of the individual probabilities, the minimum probablity required for a phosphorylation loss or gain to be called is 0.5. Any events with gain or loss probabilities less than the probability threshold are discarded. 
  	The threshold of 0.5 should suffice for calling gains and losses. However, you may wish to increase this threshold to display only events of higher confidence. 
  </p>

  <hr>
  
  
  <h4>What does the log ratio mean?</h4>
  <p>The log ratio measures how distant the wild-type score is from the mutant. A high positive log ratio represents a high confidence gain-of-phosphorylation event. 
  A high negative log ratio represents a high confidence loss-of-phosphorylation event.</p>
  
  <p>Events with a log ratio below the log ratio threshold are discarded. By default, this threshold is 1, but can be increased for in the options panel.
  
  <p><em>Note:</em> If the log ratio has a value of <code>-</code>, this means a log ratio cannot be assigned due to the fact that one of the scores has a value of 0. You can assume this log ratio is of high confidence <em>i.e.</em> approaches infinity (for gains) or -infinity (for losses)</p>

  <hr>
  
  <h4>Can I use MIMP offline?</h4>
  <p>Yes. MIMP is freely available as an R package, which can be run locally. You can find more information on how to obtain the R package on the <a target="_blank" href="/download">download page</a></p>

  <hr>
  
  <h4>I have a large amount of data, how can I speed up the process?</h4>
  <p>If you are processing a large amount of data, please download and run MIMP locally. Please refer to the <a target="_blank" href="/download">download page</a> for more information on how to download MIMP.</p>

  <hr>
  
  <h4>I can't find my kinase of interest, what should I do?</h4>
  <p>MIMP uses up-to-date data from <a target="_blank" href="http://phospho.elm.eu.org/">PhosphoELM</a>, <a target="_blank" href="http://www.phosphosite.org/">PhosphoSite Plus</a>, <a target="_blank" href="http://www.hprd.org/">HPRD</a> and <a target="_blank" href="http://phosphonetworks.org/">PhosphoNetwork</a>. If your kinase is not available it is likely that there is insufficient kinase-substrate data available to construct a specificity model. If you are able to provide sufficient kinase-substrate data for your kinase, please <a target="_blank" href="#">contact us</a>.</p>

  <hr>
  
  <h4>I'm having trouble with MIMP that isn't addressed on this page. What should I do?</h4>
  <p>If you are having trouble with MIMP please <a href="mailto:gary.bader@@utoronto.ca,juri.reimand@@utoronto.ca,wagih@@ebi.ac.uk?subject=MIMP inquiry">contact us</a> and we will try resolve it.</p>


}
